src/Authentication/Passwords/ValidationRules.php

```diff
@@ -73,7 +73,7 @@ protected function buildUserFromRequest(): User
         /** @var IncomingRequest $request */
         $request = service('request');
 
-        $data = $request->getPost($fields);
+        $data = $request->getVar($fields);
```

Comment:
To be honest, I would like to deprecate `getVar()` because it is easy to create vulnerability to bypass the validation,
and we need to consider there is no attack vectors, but it is very difficult and takes a lot of time.
We must consider all cases that attacker could send.

Since `getVar()` retrieves data from various sources (POST data, GET data, JSON data, etc), only at the runtime we know which data it actually retrieves, because an attacker may send any kind of request. The data that a developer expects to be validated (e.g. POST data) may differ from the data that is actually validated (e.g. GET data), resulting in bypassing the validation because unvalidated data is processed. 


Are you sure there is no attack vectors that bypass this validation?

Reply:
I understand your concern about security, but these are the areas I'm currently looking at:

1. If developers are asked to add a square bracket at the end of the rule like so: `strong_password[]`, just like you suggested in [this comment](https://github.com/codeigniter4/shield/issues/612#issuecomment-1411620121), it means that the docs for passwords validations would have to be updated.
2. If `getVar()` would pose some security issues, would it be better to first, check for `getPost()` and fall back to `getJSON()`?

Reply:
@sammyskills I agree with @kenjis. `getVar()` is not good idea here.

Reply:
@sammyskills Can't you make a custom rule, e.g. (`strong_password_with_getJSON`)?

Reply:
That's fine. So what do you think about my suggestion: 1 or 2?

Reply:
> @sammyskills Can't you make a custom rule, e.g. (`strong_password_with_getJSON`)?

Of course, I can create a custom rule. But let's see what @kenjis or @lonnieezell has to say about my suggestions.

Reply:
> That's fine. So what do you think about my suggestion: 1 or 2?

To be honest, I already knew that Kenjis was against using `getVar()`. But I personally like the support for `getJSON()`, and I think it's worth adding.

Reply:
Sure. 

We can still have support for JSON requests by doing something like so: 
```php
$request->getPost() ?? $request->getJSON();
```
But like I said, I need to hear from @kenjis for the best approach to take in this case. Who knows, we may even go for your suggestion of creating a new rule 😉

Reply:
I think `buildUserFromRequest()` should be removed. 
So I think using `strong_password[]` in the docs and the source code is better.

The true issue is `strong_password` guesses what data to be used for validation.
But `strong_password` never know what data should be used for validation.
It is determined by a developer (in a controller).
 In your use case, JSON data is the data to be used,
and POST data should not be used. But the following code uses POST data if exists.

```php
$request->getPost() ?? $request->getJSON();
```

All data for the validation is in JSON data in a request, and it is passed as `$data` if you use `strong_password[]`:
https://github.com/codeigniter4/shield/blob/a17a09cacffcace0cc3fa3957d2e0f23413e7470/src/Authentication/Passwords/ValidationRules.php#L31-L34
Why don't you use `$data`?

The responsibility of a validation rule is to validate passed data, not to guess what data to be used. The tight coupling of the validation rule and request object is a bad design. The problem arises because strong_password does unnecessary thing.

Reply:
> I think `buildUserFromRequest()` should be removed. So I think using `strong_password[]` in the docs and the source code is better.

Doing this _(removing `buildUserFromRequest()`)_ will result in a breaking change. All applications using the `strong_password` rule without the `[]` suffix will fail. I just tested it now.

What do you think?

Reply:
Yes, it is a breaking change. So we cannot remove it easily. I don't say do it now.

Reply:
So what do you suggest for now?

To use `strong_password[]` for JSON requests and add an explanation in the jwt docs?

Reply:
More so, since shield is still in beta, introducing a BC is not a bad idea at all.

Reply:
To use `strong_password[]` for all requests.

If you agree with removing `buildUserFromRequest()`, make it deprecated for now.

Reply:
Of course I agree.